File: src/lib/statistics.ts
import { StatisticalAnalysis, VariantResult } from '@/types';
import { analyzeABTestFromUpload } from './abTestAnalyzer';

// Calculate sample size for two-proportion z-test
export function calculateSampleSize(
  p1: number, // Control conversion rate
  p2: number, // Variant conversion rate  
  alpha: number = 0.05, // Significance level
  beta: number = 0.20, // Type II error (power = 1 - beta = 0.80)
  tailType: 'one-tailed' | 'two-tailed' = 'two-tailed'
): number {
  // Calculate critical values based on actual alpha and beta values
  const zAlpha = tailType === 'two-tailed' 
    ? normalInverseCDF(1 - alpha/2)  // Two-tailed test
    : normalInverseCDF(1 - alpha);   // One-tailed test
  const zBeta = normalInverseCDF(1 - beta); // Critical value for beta
  
  const pooledP = (p1 + p2) / 2;
  const effect = Math.abs(p2 - p1);
  
  const numerator = Math.pow(zAlpha * Math.sqrt(2 * pooledP * (1 - pooledP)) + zBeta * Math.sqrt(p1 * (1 - p1) + p2 * (1 - p2)), 2);
  const denominator = Math.pow(effect, 2);
  
  return Math.ceil(numerator / denominator);
}

// Normal inverse CDF approximation for critical values
function normalInverseCDF(p: number): number {
  if (p <= 0 || p >= 1) return 0;
  
  const c0 = 2.515517;
  const c1 = 0.802853;
  const c2 = 0.010328;
  const d1 = 1.432788;
  const d2 = 0.189269;
  const d3 = 0.001308;
  
  const t = Math.sqrt(-2 * Math.log(p > 0.5 ? 1 - p : p));
  const z = t - ((c2 * t + c1) * t + c0) / (((d3 * t + d2) * t + d1) * t + 1);
  
  return p > 0.5 ? z : -z;
}

// Calculate estimated test duration
export function calculateTestDuration(sampleSize: number, trafficPerDay: number): number {
  return Math.ceil((sampleSize * 2) / trafficPerDay); // *2 for both control and variant
}

// Perform two-proportion z-test
export function twoProportionZTest(
  x1: number, // Control conversions
  n1: number, // Control sample size
  x2: number, // Variant conversions
  n2: number, // Variant sample size
  tailType: 'one-tailed' | 'two-tailed' = 'two-tailed'
): StatisticalAnalysis {
  const p1 = x1 / n1;
  const p2 = x2 / n2;
  const pPool = (x1 + x2) / (n1 + n2);
  
  // Calculate z-score
  const standardError = Math.sqrt(pPool * (1 - pPool) * (1/n1 + 1/n2));
  const zScore = (p2 - p1) / standardError;
  
  // Calculate p-value
  const pValue = tailType === 'two-tailed' 
    ? 2 * (1 - normalCDF(Math.abs(zScore)))
    : 1 - normalCDF(zScore);
  
  // Calculate confidence interval for difference
  const seDiff = Math.sqrt((p1 * (1 - p1) / n1) + (p2 * (1 - p2) / n2));
  const marginOfError = 1.96 * seDiff; // 95% confidence
  const diff = p2 - p1;
  
  const confidenceInterval = {
    lower: diff - marginOfError,
    upper: diff + marginOfError
  };
  
  // Calculate uplift
  const uplift = ((p2 - p1) / p1) * 100;
  
  return {
    sampleSize: n1 + n2,
    pValue,
    confidenceInterval,
    uplift,
    isSignificant: pValue < 0.05
  };
}

// Normal CDF approximation
function normalCDF(x: number): number {
  const t = 1 / (1 + 0.2316419 * Math.abs(x));
  const d = 0.3989423 * Math.exp(-x * x / 2);
  const probability = d * t * (0.3193815 + t * (-0.3565638 + t * (1.781478 + t * (-1.821256 + t * 1.330274))));
  
  return x > 0 ? 1 - probability : probability;
}

// Calculate confidence interval for a proportion
export function calculateConfidenceInterval(
  conversions: number,
  sampleSize: number,
  confidenceLevel: number = 0.95
): { lower: number; upper: number } {
  const p = conversions / sampleSize;
  const z = confidenceLevel === 0.95 ? 1.96 : 1.645;
  const standardError = Math.sqrt((p * (1 - p)) / sampleSize);
  const marginOfError = z * standardError;
  
  return {
    lower: Math.max(0, p - marginOfError),
    upper: Math.min(1, p + marginOfError)
  };
}

// Analyze A/B test data using the new analyzer
export async function analyzeTestData(
  data: any[][],
  variantColumn: number,
  conversionColumn: number,
  columns: string[],
  statisticalParams?: { 
    significanceLevel?: number; 
    power?: number; 
    confidenceLevel?: number;
  }
): Promise<{ variants: VariantResult[]; analysis: StatisticalAnalysis | null }> {
  try {
    const variantColumnName = columns[variantColumn];
    const conversionColumnName = columns[conversionColumn];
    
    // First check if we have continuous data (non-binary values in conversion column)
    const conversionValues = data.slice(1).map(row => row[conversionColumn]).filter(val => val !== null && val !== undefined);
    const isContinuous = conversionValues.some(val => {
      const num = parseFloat(val);
      return !isNaN(num) && num !== 0 && num !== 1;
    });
    
    if (isContinuous) {
      // Handle continuous data
      const variantGroups: { [key: string]: number[] } = {};
      
      data.slice(1).forEach(row => {
        const variant = row[variantColumn];
        const value = parseFloat(row[conversionColumn]);
        
        if (!isNaN(value)) {
          if (!variantGroups[variant]) {
            variantGroups[variant] = [];
          }
          variantGroups[variant].push(value);
        }
      });
      
      const variantNames = Object.keys(variantGroups);
      if (variantNames.length < 2) {
        throw new Error('Need at least 2 variants for comparison');
      }
      
      // Prepare variants data
      const variants: VariantResult[] = variantNames.map(name => ({
        name,
        visitors: variantGroups[name].length,
        conversions: 0, // Not applicable for continuous data
        conversionRate: 0, // Not applicable for continuous data
        confidenceInterval: { lower: 0, upper: 0 },
        continuousValues: variantGroups[name]
      }));
      
      // Basic analysis for continuous data
      const analysis: StatisticalAnalysis = {
        sampleSize: variants.reduce((sum, v) => sum + v.visitors, 0),
        pValue: 0.5, // Default for continuous data without advanced analysis
        confidenceInterval: { lower: 0, upper: 0 },
        uplift: 0, // Calculate based on means for continuous data
        isSignificant: false
      };
      
      return { variants, analysis };
    } else {
      // Handle categorical/binary data (existing logic)
      const results = analyzeABTestFromUpload(data, variantColumnName, conversionColumnName, columns);
      
      // Convert to the expected format
      const variants: VariantResult[] = [
        // Add control first
        {
          name: results.control.variant,
          visitors: results.control.visitors,
          conversions: results.control.conversions,
          conversionRate: results.control.conversion_rate / 100,
          confidenceInterval: calculateConfidenceInterval(results.control.conversions, results.control.visitors)
        },
        // Add other variants
        ...results.results.map(result => ({
          name: result.variant,
          visitors: result.visitors,
          conversions: result.conversions,
          conversionRate: result.conversion_rate / 100,
          confidenceInterval: calculateConfidenceInterval(result.conversions, result.visitors)
        }))
      ];
      
      // Use standard two-proportion z-test for analysis
      const control = variants.find(v => v.name.toLowerCase().includes('control')) || variants[0];
      const treatment = variants.find(v => !v.name.toLowerCase().includes('control')) || variants[1];
      
      let analysis: StatisticalAnalysis;
      if (control && treatment && variants.length >= 2) {
        analysis = twoProportionZTest(
          control.conversions,
          control.visitors,
          treatment.conversions,
          treatment.visitors
        );
      } else {
        analysis = {
          sampleSize: variants.reduce((sum, v) => sum + v.visitors, 0),
          pValue: 0.5,
          confidenceInterval: { lower: 0, upper: 0 },
          uplift: results.results[0]?.uplift || 0,
          isSignificant: false
        };
      }
      
      return { variants, analysis };
    }
  } catch (error) {
    console.error('Analysis error:', error);
    
    // Fallback to original analysis
    const variantGroups: { [key: string]: any[] } = {};
    
    data.forEach(row => {
      const variant = row[variantColumn];
      if (!variantGroups[variant]) {
        variantGroups[variant] = [];
      }
      variantGroups[variant].push(row);
    });
    
    const variants: VariantResult[] = Object.entries(variantGroups).map(([name, rows]) => {
      const visitors = rows.length;
      const conversions = rows.filter(row => {
        const conversionValue = row[conversionColumn];
        return conversionValue === 'Yes' || conversionValue === '1' || conversionValue === 1 || conversionValue === true;
      }).length;
      
      const conversionRate = visitors > 0 ? conversions / visitors : 0;
      const confidenceInterval = calculateConfidenceInterval(conversions, visitors);
      
      return {
        name,
        visitors,
        conversions,
        conversionRate,
        confidenceInterval
      };
    });
    
    let analysis: StatisticalAnalysis | null = null;
    const control = variants.find(v => v.name.toLowerCase().includes('control'));
    const variant = variants.find(v => !v.name.toLowerCase().includes('control'));
    
    if (control && variant) {
      analysis = twoProportionZTest(
        control.conversions,
        control.visitors,
        variant.conversions,
        variant.visitors
      );
    }
    
    return { variants, analysis };
  }
}

File: src/lib/statistics.py
# """
# Python equivalent of statistics.ts
# Statistical functions for A/B testing including sample size calculation, z-tests, and confidence intervals
# """

import numpy as np
import scipy.stats as stats
from typing import List, Dict, Any, Optional, Tuple, Union


def calculate_sample_size(
    p1: float,  # Control conversion rate
    p2: float,  # Variant conversion rate
    alpha: float = 0.05,  # Significance level
    beta: float = 0.20,  # Type II error (power = 1 - beta = 0.80)
    tail_type: str = 'two-tailed'
) -> int:
    """Calculate sample size for two-proportion z-test"""
    # Calculate critical values based on actual alpha and beta values
    if tail_type == 'two-tailed':
        z_alpha = stats.norm.ppf(1 - alpha/2)
    else:
        z_alpha = stats.norm.ppf(1 - alpha)
    
    z_beta = stats.norm.ppf(1 - beta)
    
    pooled_p = (p1 + p2) / 2
    effect = abs(p2 - p1)
    
    numerator = (z_alpha * np.sqrt(2 * pooled_p * (1 - pooled_p)) + 
                z_beta * np.sqrt(p1 * (1 - p1) + p2 * (1 - p2)))**2
    denominator = effect**2
    
    return int(np.ceil(numerator / denominator))


def calculate_test_duration(sample_size: int, traffic_per_day: int) -> int:
    """Calculate estimated test duration"""
    return int(np.ceil((sample_size * 2) / traffic_per_day))  # *2 for both control and variant


def two_proportion_z_test(
    x1: int,  # Control conversions
    n1: int,  # Control sample size
    x2: int,  # Variant conversions
    n2: int,  # Variant sample size
    tail_type: str = 'two-tailed'
) -> Dict[str, Any]:
    """Perform two-proportion z-test"""
    p1 = x1 / n1
    p2 = x2 / n2
    p_pool = (x1 + x2) / (n1 + n2)
    
    # Calculate z-score
    standard_error = np.sqrt(p_pool * (1 - p_pool) * (1/n1 + 1/n2))
    z_score = (p2 - p1) / standard_error
    
    # Calculate p-value
    if tail_type == 'two-tailed':
        p_value = 2 * (1 - stats.norm.cdf(abs(z_score)))
    else:
        p_value = 1 - stats.norm.cdf(z_score)
    
    # Calculate confidence interval for difference
    se_diff = np.sqrt((p1 * (1 - p1) / n1) + (p2 * (1 - p2) / n2))
    margin_of_error = 1.96 * se_diff  # 95% confidence
    diff = p2 - p1
    
    confidence_interval = {
        'lower': diff - margin_of_error,
        'upper': diff + margin_of_error
    }
    
    # Calculate uplift
    uplift = ((p2 - p1) / p1) * 100 if p1 != 0 else 0
    
    return {
        'sample_size': n1 + n2,
        'p_value': p_value,
        'confidence_interval': confidence_interval,
        'uplift': uplift,
        'is_significant': p_value < 0.05
    }


def calculate_confidence_interval(
    conversions: int,
    sample_size: int,
    confidence_level: float = 0.95
) -> Dict[str, float]:
    """Calculate confidence interval for a proportion"""
    p = conversions / sample_size
    z = 1.96 if confidence_level == 0.95 else 1.645
    standard_error = np.sqrt((p * (1 - p)) / sample_size)
    margin_of_error = z * standard_error
    
    return {
        'lower': max(0, p - margin_of_error),
        'upper': min(1, p + margin_of_error)
    }


def analyze_test_data(
    data: List[List[Any]],
    variant_column: int,
    conversion_column: int,
    columns: List[str],
    statistical_params: Optional[Dict[str, float]] = None
) -> Dict[str, Any]:
    """Analyze A/B test data using statistical analysis"""
    try:
        variant_column_name = columns[variant_column]
        conversion_column_name = columns[conversion_column]
        
        # Check if we have continuous data
        conversion_values = [row[conversion_column] for row in data[1:] 
                           if row[conversion_column] is not None]
        is_continuous = any(
            isinstance(val, (int, float)) and val != 0 and val != 1 
            for val in conversion_values
        )
        
        if is_continuous:
            # Handle continuous data
            variant_groups = {}
            
            for row in data[1:]:
                variant = row[variant_column]
                try:
                    value = float(row[conversion_column])
                    if not np.isnan(value):
                        if variant not in variant_groups:
                            variant_groups[variant] = []
                        variant_groups[variant].append(value)
                except (ValueError, TypeError):
                    continue
            
            variant_names = list(variant_groups.keys())
            if len(variant_names) < 2:
                raise ValueError('Need at least 2 variants for comparison')
            
            # Prepare variants data
            variants = []
            for name in variant_names:
                variants.append({
                    'name': name,
                    'visitors': len(variant_groups[name]),
                    'conversions': 0,  # Not applicable for continuous data
                    'conversion_rate': 0,  # Not applicable for continuous data
                    'confidence_interval': {'lower': 0, 'upper': 0},
                    'continuous_values': variant_groups[name]
                })
            
            # Perform t-test for continuous data
            if len(variant_names) >= 2:
                group_a = variant_groups[variant_names[0]]
                group_b = variant_groups[variant_names[1]]
                
                t_stat, p_value = stats.ttest_ind(group_a, group_b, equal_var=False)
                
                mean_diff = np.mean(group_a) - np.mean(group_b)
                pooled_std = np.sqrt((np.var(group_a, ddof=1) + np.var(group_b, ddof=1)) / 2)
                cohens_d = mean_diff / pooled_std if pooled_std != 0 else 0
                
                # Confidence interval for difference
                se = np.sqrt(np.var(group_a, ddof=1)/len(group_a) + np.var(group_b, ddof=1)/len(group_b))
                df = min(len(group_a), len(group_b)) - 1
                t_critical = stats.t.ppf(0.975, df)
                margin_of_error = t_critical * se
                
                analysis = {
                    'sample_size': sum(len(variant_groups[name]) for name in variant_names),
                    'p_value': p_value,
                    'confidence_interval': {
                        'lower': mean_diff - margin_of_error,
                        'upper': mean_diff + margin_of_error
                    },
                    'uplift': 0,  # Calculate based on means for continuous data
                    'is_significant': p_value < (statistical_params.get('significance_level', 0.05) if statistical_params else 0.05)
                }
            else:
                analysis = None
            
            return {'variants': variants, 'analysis': analysis}
        
        else:
            # Handle categorical/binary data
            variant_groups = {}
            
            for row in data:
                variant = row[variant_column]
                if variant not in variant_groups:
                    variant_groups[variant] = []
                variant_groups[variant].append(row)
            
            variants = []
            for name, rows in variant_groups.items():
                visitors = len(rows)
                conversions = sum(1 for row in rows 
                                if row[conversion_column] in ['Yes', '1', 1, True])
                
                conversion_rate = conversions / visitors if visitors > 0 else 0
                confidence_interval = calculate_confidence_interval(conversions, visitors)
                
                variants.append({
                    'name': name,
                    'visitors': visitors,
                    'conversions': conversions,
                    'conversion_rate': conversion_rate,
                    'confidence_interval': confidence_interval
                })
            
            # Perform z-test for binary data
            analysis = None
            control = next((v for v in variants if 'control' in v['name'].lower()), None)
            variant = next((v for v in variants if 'control' not in v['name'].lower()), None)
            
            if control and variant:
                analysis = two_proportion_z_test(
                    control['conversions'],
                    control['visitors'],
                    variant['conversions'],
                    variant['visitors']
                )
            
            return {'variants': variants, 'analysis': analysis}
    
    except Exception as e:
        print(f"Analysis error: {e}")
        # Fallback analysis would go here
        return {'variants': [], 'analysis': None}

# """